CARET_trace/src/tracing_controller.cpp

```diff
@@ -247,7 +247,7 @@ bool TracingController::is_allowed_callback(const void * callback)
 
 bool TracingController::is_allowed_node(const void * node_handle)
 {
-  std::lock_guard<std::shared_timed_mutex> lock(mutex_);
+  std::shared_lock<std::shared_timed_mutex> lock(mutex_);  // read lock
```

Comment:
Next `auto node_name = node_handle_to_node_names_[node_handle];` modify `node_handle_to_node_names_` when it does not have this key, so shared_lock cause problem. I suggest changing lines after shared_lock like below:

```
  auto node_name_it = node_handle_to_node_names_.find(node_handle);
  if (node_name_it == node_handle_to_node_names_.end()) {
    return true;
  }
  if (select_enabled_ && selected_node_names_.size() > 0) {
    auto is_selected_node = partial_match(selected_node_names_, node_name_it.second);
    return is_selected_node;
  } else if (ignore_enabled_ && ignored_node_names_.size() > 0) {
    auto is_ignored_node = partial_match(ignored_node_names_, node_name_it.second);
    return !is_ignored_node;
  }
  return true;
```

Reply:
I see, if the key does not exist, the writing will occur.